# Deep Neural Networks for Time Series Forecasting the Easy Way
## Getting the Data from the Internet
如果你曾在新加坡生活过，你就会知道，任何人想要注册新车，都必须先获得权利证书（COE）。它使持有人有权拥有车辆，并在这个热带岛国非常有限的道路上行驶。发放的特遣队所属装备数量有限，只能通过公开招标制度获得。

In [28]:
import numpy as np
import pandas as pd

In [39]:
loc = "COE.xls"

## Cleaning up Downloaded Spreadsheet Files

In [40]:
Excel_file = pd.ExcelFile(loc)

In [41]:
print(Excel_file.sheet_names)

['COE data']


In [42]:
spreadsheet = Excel_file.parse('COE data')

In [32]:
spreadsheet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 265 entries, 0 to 264
Data columns (total 6 columns):
DATE      265 non-null datetime64[ns]
COE$      265 non-null float64
COE$_1    265 non-null float64
#Bids     265 non-null int64
Quota     265 non-null int64
Open?     265 non-null int64
dtypes: datetime64[ns](1), float64(2), int64(3)
memory usage: 12.5 KB


我们看到电子表格包含六列，有265行（条目）。第一栏包含COE拍卖的日期，其余各栏为其他五个变量。变量COE$包含权利证书的历史价格。我们将把它作为我们的目标变量，并把它存储在一个名为data的对象中。

In [33]:
data = spreadsheet['COE$']
print(data.head())

0     7400.0
1    11100.0
2     5002.0
3     3170.0
4     3410.0
Name: COE$, dtype: float64


## Adjusting Dates


In [43]:
print(spreadsheet['DATE'][193:204])

193   2004-02-01
194   2002-02-15
195   2004-03-01
196   2004-03-15
197   2004-04-01
198   2002-04-15
199   2004-05-01
200   2004-05-15
201   2004-06-01
202   2002-06-15
203   2004-07-01
Name: DATE, dtype: datetime64[ns]


In [44]:
spreadsheet.loc[194,'DATE']='2004-02-05'
spreadsheet.loc[198,'DATE']='2004-04-15'
spreadsheet.loc[202,'DATE']='2004-06-15'

In [45]:
print(spreadsheet['DATE'][193:204])

193    2004-02-01 00:00:00
194             2004-02-05
195    2004-03-01 00:00:00
196    2004-03-15 00:00:00
197    2004-04-01 00:00:00
198             2004-04-15
199    2004-05-01 00:00:00
200    2004-05-15 00:00:00
201    2004-06-01 00:00:00
202             2004-06-15
203    2004-07-01 00:00:00
Name: DATE, dtype: object


## Saving Data

In [46]:
spreadsheet.to_csv('update_COE.csv')

神经元的任务是对输入信号进行加权和，并在将输出传给下一层之前应用激活函数。所以，我们看到，输入层将数据传递给第一个隐藏层。隐藏层神经元对输入层神经元传递给他们的信息进行求和；输出层神经元对隐藏层神经元传递给他们的加权信息进行求和。
## How to Scale the Input attributes
深度神经网络对输入数据的尺度很敏感，特别是在使用sigmoid（见第38页）或tanh（见第44页）激活函数时。最好的做法是将数据重新标定到-1到+1或0到1的范围内，这可以通过sklearn模块中的MinMaxScaler方法实现。

In [47]:
from sklearn import preprocessing

In [48]:
x = data
scaler = preprocessing.MinMaxScaler(feature_range=(0,1))
# 第一行将价格数据传输到我们的属性变量x中，
# 然后加载预处理模块，并使用MinMaxScaler与feature_range设置，
# 将属性缩放到0到1的范围。对象缩放器，包含缩放数据所需的指令。

In [49]:
scaler

MinMaxScaler(copy=True, feature_range=(0, 1))

这告诉我们scaler将把数据放大到0,1范围。请注意，复制=True是默认设置。如果输入已经是一个numpy数组，你可以将其设置为False，以执行原地行归一化，避免复制。但是，我们将其设置为True，因为x（和数据）是Pandas系列的。

Pandas对于数据操作来说是很好的，但是对于我们的数值分析来说，我们需要使用一个numpy数组。

In [50]:
x = np.array(x).reshape(( len(x), ))

## Log Transform
当处理总是取正值的时间序列时，我通常喜欢对数据进行对数转换。我想这源于我在建立计量经济学模型的日子里，应用对数转换有助于数据的标准化。总之，下面是如何做到这一点。

In [51]:
x = np.log(x)